src/gl/implementation/gl-attributes.adb

```diff
@@ -53,6 +53,102 @@ package body GL.Attributes is
+                                       Stride * Short'Size / System.Storage_Unit,
+                                       Offset * Short'Size / System.Storage_Unit);
+      end case;
+      Raise_Exception_On_OpenGL_Error;
+   end Set_Vertex_Attrib_Pointer;
+
+   procedure Set_Vertex_Attrib_Pointer2 (Index          : Attribute;
+                                         Count          : Component_Count;
+                                         Kind           : Numeric_Type;
+                                         Stride_In_Bytes, Offset_In_Bytes : Size) is
+   begin
+      case Kind is
+         when Single_Type =>
+            API.Vertex_Attrib_Pointer (Index, Count, Kind, Low_Level.False,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when Double_Type =>
+            API.Vertex_AttribL_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when UInt_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when UByte_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when UShort_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when Int_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when Byte_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+         when Short_Type =>
+            API.Vertex_AttribI_Pointer (Index, Count, Kind,
+                                       Stride_In_Bytes,
+                                       Offset_In_Bytes);
+      end case;
+      Raise_Exception_On_OpenGL_Error;
+   end Set_Vertex_Attrib_Pointer2;
+
+   procedure Set_Vertex_Attrib_Pointer2 (Index          : Attribute;
+                                         Count          : Component_Count;
+                                         Kind           : Numeric_Type;
+                                         Normalized     : Boolean;
+                                         Stride_In_Bytes, Offset_In_Bytes : Size) is
+   begin
+      API.Vertex_Attrib_Pointer (Index, Count, Kind,
+                                 Low_Level.Bool (Normalized),
+                                 Stride_In_Bytes, Offset_In_Bytes);
+      Raise_Exception_On_OpenGL_Error;
+   end Set_Vertex_Attrib_Pointer2;
```

Comment:
I think having `Stride` and `Offset` depending on the data type was a bad idea (there is no reason why `Stride` should be a multiple of the type's size), moreover automatic matching to the underlying function also was a bad idea (there's nothing wrong with giving integer values and specifying that they should be converted to floats). And I think the best thing to do is to fix this properly. The name `Set_Vertex_Attrib_Pointer2` is not helping the user understand what's going on. I suggest the following:

 * Keep the original `Set_Vertex_Attrib_Pointer` to not break existing code, comment its deprecation in the `.ads` (sadly, `pragma Obsolescent` is not standard).
 * Add `Set_Vertex_Float_Attrib_Pointer`, `Set_Vertex_Integer_Attrib_Pointer` and `Set_Vertex_Double_Attrib_Pointer` to match the three underlying API functions, and for each of them, use byte offsets and strides.
 * Update the `.spec` file so that those API functions point to their proper wrappers.

This should give you the same set of possibilities as your current solution.

Reply:
According to spec https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glVertexAttribPointer.xhtml 
and especially that :

> type:
Specifies the data type of each component in the array. The symbolic constants GL_BYTE, GL_UNSIGNED_BYTE, GL_SHORT, GL_UNSIGNED_SHORT, GL_INT, and GL_UNSIGNED_INT are accepted by glVertexAttribPointer and glVertexAttribIPointer. Additionally GL_HALF_FLOAT, GL_FLOAT, GL_DOUBLE, GL_FIXED, GL_INT_2_10_10_10_REV, GL_UNSIGNED_INT_2_10_10_10_REV and GL_UNSIGNED_INT_10F_11F_11F_REV are accepted by glVertexAttribPointer. GL_DOUBLE is also accepted by glVertexAttribLPointer and is the only token accepted by the type parameter for that function.

I decided to do Set_Vertex_Attrib_Pointer with full declaration, keep the original function with warning massage, and add Set_Vertex_Integer_Attrib_Pointer and Set_Vertex_Double_Attrib_Pointer as you advised.
What do you say?